Add snabb config derive-data-format program

tsyesika opened the pull request:
This introduces the `snabb config` program. The program only has one option at the moment, that option being `derive-data-format` which takes one argument, a yang schema module path. The program will read a schema, derive the data grammar and then display in a human readable way how the data configuration should look (see more in #529).

An example usage of this would be:
**simple-router.yang**
```yang
module snabb-simple-router {
  namespace snabb:simple-router;
  prefix simple-router;

  import ietf-inet-types {prefix inet;}

  leaf active { type boolean; default true; }

  container routes {
    presence true;
    list route {
      key addr;
      leaf addr { type inet:ipv4-address; mandatory true; }
      leaf port { type uint8 { range 0..11; } mandatory true; }
    }
  }
}
```

**Command**
```sh
$ sudo ./src/snabb config derive-data-format ./src/lib/yang/simple-router.yang
routes {
    // List: this nested structure is repeated with (a) unique key(s)
    route {
        addr inet:ipv4-address; // mandatory key
        port uint8; // mandatory between 0..11
    }
}
active boolean;
```

PTAL @wingo

tsyesika committed:
Add `snabb config derive-data-format` option

This introduces the config program and adds an option to produce an
example of what the data format should look like based upon a valid
YANG schema.

wingo commented:
Would you mind moving this code to `program/config/describe_data_format/describe_data_format.lua` and add a `program/config/config.lua` that looks like `program/lwaftr/lwaftr.lua`, with usage information as README files as the lwaftr does?  Thanks :)

tsyesika committed:
Move data_format into own directory and rename

This relocates the command "derive-data-format" into it's own directory
to make it cleaner and easier to add new commands to the `snabb config`
program. This also renames it from "derive-data-format" to simply be
"data-format".

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,119 @@
+      end
+   end
+
+   if #args <= 0 then
```
Would you mind refactoring to use `lib.getopt` and add a --help argument?  Like this:

``` lua
local lib = require('core.lib')

local function show_usage(code)
   print(require("program.config.data_format.README_inc"))
   main.exit(code)
end

local function parse_args(args)
   local handlers = {}
   function handlers.h() show_usage(0) end
   args = lib.dogetopt(args, handlers, "h", { help="h" })
   if #args ~= 1 then show_usage(1) end
   return unpack(args)
end

local function display_data_format(...)
   ...
end

function run(args)
   local schema_file_name = parse_args(args)
   local grammar = ...
   display_data_format(grammar, 0)
   main.exit(0)
end
```

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,119 @@
+	 print_level(level, name.." {")
+         display_data_format(node, level + 1)
+         print_level(level, "}")
+      elseif node.type == "scalar" then
```
Here instead of the if/elseif chain, what do you think about writing this in a data-driven way?  Like this:

``` lua
local describers = {}
local function describe(keyword, production)
   assert(describers[production.type])(keyword, production)
end
local function describe_sequence(productions)
   local keywords = {} -- sort keys of productions into this list
   for _,keyword in ipairs(keywords) do describe(keyword, productions[keyword]) end
end
function describers.scalar(keyword, production)
   -- here we describe this production...
end
function describers.table(keyword, production) ... end
-- for array and struct as well

--- for top level which is a struct, result is:
describe_sequence(production.members)
```

tsyesika committed:
Use core.lib.dogetopt to support -h and --help

This adds flag support for both `snabb config` and the
`snabb config data-format` sub-program. They now both respond
appropriately to the "-h" and "--help" flags by displaying the
appropriate help information.

tsyesika committed:
Convert node displaying into data driven code

This converts a if-elseif-else block which was being used
as a mechanism for selecting between different node types
and instead converting it to a more function driven method.

Now there is a table of methods built up that handle the different
types and those are selected based of the node type to handle the
display of the data.

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,135 @@
+local lib = require("core.lib")
+local schema = require("lib.yang.schema")
+local yang_data = require("lib.yang.data")
+local usage = require("program.config.data_format.README_inc")
```
I think we can remove this require, as it's required in place below.

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,135 @@
+local usage = require("program.config.data_format.README_inc")
+
+-- Number of spaces a tab should consist of when indenting config.
+local tab_spaces = 4
```
We should use the same as what the serializer produces, which is currently 2.  Can you use 2 here?

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,135 @@
+end
+
+function run(args)
+   function print_level(level, ...)
```
Because you don't prefix these with "local", all these functions (except the ones of the form `function foo.bar`) are getting added to the top level.  But actually I think that's where they should be -- `run` should handle parsing arguments and exiting successfully but other things should be done outside run.  Would you mind making that refactor please?

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,135 @@
+	    if name == k then options.key = true end
+	 end
+      end
+      if node.argument_type then
```
I think it's not the right thing to collate options on all productions.  There are only 4 production kinds: table, array, struct, and scalar.  The set of properties of interest on these productions is fixed.  E.g. argument_type is only a thing on a scalar production.  keys is only a property on tables, mandatory is only on scalars, etc.  Can you push these things down into the handlers for the various kinds please?

wingo reviewed (changes requested):
Looking good, though I have some requests to make please :)  Thank you in advance for humoring me and let me know if any of these concerns are off base!

tsyesika committed:
Refactor options and data_format config file

This whilst refactoring most of data_format is largely to add refactor
options so that they are only calculated when the option can actually
be set. This also makes some small changes like removing an unused
require statement and changing the tab width from 4 to 2

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,155 @@
+   local handlers = {}
+   handlers.h = function() show_usage(0) end
+   args = lib.dogetopt(args, handlers, "h", {help="h"})
+   if #args <= 0 then show_usage(1) end
```
#args can't be negative; shouldn't this be `~= 1`?

tsyesika replied:
If anything it should be `~= 0`.

`snabb config data-format` is called correctly with two arguments. One of which being the sub-program itself `data-format` and the other being the path name. `config.lua` during it's sub-program check and dispatch removes the first element (see https://github.com/Igalia/snabb/blob/config-program/src/program/config/config.lua#L21) which leaves the arguments as `{1 = "path to schema file"}`. We're then checking if that is missing so when `data-format` is called without that `args` will be of length 0.

wingo replied:
this should error: `snabb config data-format foo bar baz`.  so we need to show usage if #args is `<` 1 and also `>` 1 which means ~= 1.  no?

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,155 @@
+end
+
+function describers.array(level, name, node)
+   print_level(
```
can you tighten up these parens?  i think the normal style is just `f(arg` unless `arg` is too long in which case the next line can start with a 3-char indent, but in no case should there be a lonely closing paren

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,155 @@
+end
+
+function describers.table(level, name, node)
+   if node.keys then
```
node.keys is always true

wingo commented on src/program/config/data_format/data_format.lua:
```diff
@@ -0,0 +1,155 @@
+function describers.scalar(level, name, node, is_key)
+   local opts = {}
+   if is_key then opts.key = is_key(name, node) end
+   opts.mandatory = options.mandatory(name, node)
```
just access node.mandatory, either it's true or falsey (false if given or nil if not)

wingo reviewed (approved):
A couple small nits, otherwise lgtm!

tsyesika committed:
Fix some small code cosmeic issues

xray7224 commented:
Fixed those nits.

kbara commented:
@xray7224 Mind deleting this branch?
